src/network/network_server.cpp

```diff
@@ -1813,12 +1791,63 @@ void NetworkServer_Tick(bool send_frame)
 	}
 }
 
+/** Helper function to restart the map. */
+static void RestartMapHelper()
+{
+	_settings_newgame.game_creation.generation_seed = GENERATE_NEW_SEED;
+	switch (_file_to_saveload.abstract_ftype) {
+		case FT_SAVEGAME:
+		case FT_SCENARIO:
+			_switch_mode = SM_LOAD_GAME;
+			break;
+
+		case FT_HEIGHTMAP:
+			_switch_mode = SM_START_HEIGHTMAP;
+			break;
+
+		default:
+			_switch_mode = SM_NEWGAME;
+	}
+}
+
+/** Check if we want to restart the map based on the year. */
+static void NetworkCheckRestartMapYear()
+{
+	/* If setting is 0, this feature is disabled. */
+	if (_settings_client.network.restart_game_year == 0) return;
+
+	if (TimerGameCalendar::year >= _settings_client.network.restart_game_year) {
+		Debug(net, 3, "Auto-restarting map: year {} reached", TimerGameCalendar::year);
+		RestartMapHelper();
+	}
+}
+
+/** Timer to restart a network server automatically based on real-time hours played. Initialized at zero to disable until settings are loaded. */
+static IntervalTimer<TimerGameRealtime> _network_restart_map_timer({std::chrono::milliseconds::zero(), TimerGameRealtime::UNPAUSED}, [](auto)
+{
+	if (!_network_server) return;
+
+	/* If setting is 0, this feature is disabled. */
+	if (_settings_client.network.restart_hours == 0) return;
+
+	Debug(net, 3, "Auto-restarting map: {} hours played", _settings_client.network.restart_hours);
+	RestartMapHelper();
+});
+
+/** Reset the automatic network restart time interval. */
+void ChangeNetworkRestartTime()
+{
+	if (!_network_server) return;
+
+	_network_restart_map_timer.SetInterval({ std::chrono::hours(_settings_client.network.restart_hours), TimerGameRealtime::UNPAUSED }, true);
```

Comment:
The part I am doubting, is the `true` part. This means if you change the setting while a game is running, the counter resets. I am not sure that is the expected behaviour. But as you reuse this function to reset the timer on game-start, it is kinda tricky.

Maybe send the `bool reset` as parameter, and reset it when you start a new game, but not when you change the setting. And yes, that can mean if you change the setting the game all of a sudden resets :P This is where I am not sure if my idea works. But if we document the setting as "time since start", that would kinda make sense.

Now I think about it, the `year` based setting has the same. When you change it just before year change, it will directly reset. So I guess indeed, not resetting is the better thing to do when changing the setting. I think :P